File: app/ecs/core/EntityManager.lua
local ____exports = {}
function ____exports.EntityManager()
    local UIDsEntity = 0
    local entities = {}
    local function newEntity()
        local id = UIDsEntity
        if entities[id] == nil then
            entities[id] = {}
        end
        UIDsEntity = UIDsEntity + 1
        return id
    end
    local function getEntity(id)
        return entities[id]
    end
    local function addComponent(entityId, component)
        entities[entityId][component.UUID] = component
    end
    return {newEntity = newEntity, getEntity = getEntity, addComponent = addComponent}
end
return ____exports

File: src/ecs/core/EntityManager.ts
import { ComponentsIDs, Components, Component } from "../baseComponents/Components";


type EntityComponents = { [key in ComponentsIDs[number]]: Components[key] };
export function EntityManager() {

    let UIDsEntity = 0;
    const entities: { [id: number]: EntityComponents } = {};

    function newEntity() {
        const id = UIDsEntity;
        if (entities[id] == undefined)
            entities[id] = {} as EntityComponents;

        UIDsEntity += 1;
        return id;
    }

    function getEntity(id: number) {
        return entities[id];
    }

    function addComponent<compID extends ComponentsIDs[number]>(entityId: number, component: Component<compID>) {
        (entities[entityId][component.UUID] as any) = component;
    }

    

    return {
        newEntity, getEntity, addComponent
    };

}